Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp

```diff
@@ -0,0 +1,168 @@
+
+    // FIXME: Support view formats
+    if (!configuration.viewFormats.isEmpty())
+        return false;
+
+    return true;
+}
+
+RefPtr<WebGPUSurfaceCocoa> WebGPUSurfaceCocoa::create(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+{
+    if (!validateSurfaceCreation(size, configuration))
+        return nullptr;
+
+    return adoptRef(*new WebGPUSurfaceCocoa(WTFMove(size), WTFMove(configuration)));
+}
+
+WebGPUSurfaceCocoa::WebGPUSurfaceCocoa(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+    : m_size(WTFMove(size))
+    , m_configuration(WTFMove(configuration))
+{
+    ASSERT(validateSurfaceCreation(size, configuration));
+}
+
+void WebGPUSurfaceCocoa::prepareToDelegateDisplay(PlatformCALayer&)
+{
+}
+
+void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)
+{
+    if (m_frontBuffer)
+        layer.setContents(m_frontBuffer->surface());
+}
+
+GraphicsLayer::CompositingCoordinatesOrientation WebGPUSurfaceCocoa::orientation() const
+{
+    return GraphicsLayer::CompositingCoordinatesOrientation::TopDown;
+}
+
+void WebGPUSurfaceCocoa::present()
+{
+    std::swap(m_frontBuffer, m_backBuffer);
+    m_backBufferTexture = nullptr;
+    std::swap(m_backBuffer, m_secondaryBackBuffer);
+}
+
+bool WebGPUSurfaceCocoa::isValidBuffer(const IOSurface* buffer)
+{
+    if (!buffer)
+        return false;
+
+    if (buffer->size() != m_size)
+        return false;
+
+    if (buffer->isInUse())
+        return false;
+
+    return true;
+}
+
+void WebGPUSurfaceCocoa::ensureValidBackBuffer()
+{
+    if (isValidBuffer(m_backBuffer.get()))
+        return;
+
+    // The back buffer is not valid, so we should not reuse it and the back buffer texture.
+    // Destroy the back buffer texture here.
+    if (m_backBufferTexture) {
+        m_backBufferTexture->destroy();
+        m_backBufferTexture = nullptr;
+    }
+
+    // Then try to recycle the secondary back buffer as the new back buffer.
+    m_backBuffer = WTFMove(m_secondaryBackBuffer);
+
+    if (isValidBuffer(m_backBuffer.get()))
+        return;
+
+    // Can't use the secondary back buffer, so destroy it and create a new one.
+    m_backBuffer = IOSurface::create(nullptr, m_size, colorSpace());
+    ASSERT(isValidBuffer(m_backBuffer.get()));
+}
+
+Ref<PAL::WebGPU::Texture> WebGPUSurfaceCocoa::currentTexture()
+{
+    ensureValidBackBuffer();
+
+    if (!m_backBufferTexture) {
+        PAL::WebGPU::TextureDescriptor textureDescriptor {
+            { "WebGPU Display texture"_s },
+            PAL::WebGPU::Extent3DDict { static_cast<unsigned>(m_size.width()), static_cast<unsigned>(m_size.height()), 1 },
+            1 /* mipMapCount */,
+            1 /* sampleCount */,
+            PAL::WebGPU::TextureDimension::_2d,
+            m_configuration.format,
+            m_configuration.usage,
+            m_configuration.viewFormats
+        };
+
+        m_backBufferTexture = m_configuration.device.createIOSurfaceBackedTexture(textureDescriptor, m_backBuffer->surface());
```

Comment:
I think this is a UAF. m_configuration doesn't retain the device.

Reply:
I'm also not super sure this is the right design. I kind of feel:
1. The Surface is conceptually the equivalent of a `CAMetalLayer`. It shouldn't need to retain anything WebGPU-specific inside it.
2. The Surface owns 3 IOSurfaces, so that part matches CAMetalLayer. So that's good
3. But the 3 IOSurfaces are created lazily, and sent to the GPU Process one-by-one, which I don't think makes a lot of sense. I think it's a better design to package up all 3 IOSurfaces and send them in a single package to the GPUP (perhaps inside GPUCanvasContext.configure(), or maybe perhaps earlier).
4. I'm also not sure that this belongs in platform/. This class is about interfacing between canvas, the compositor, and webgpu. 
5. I'm also not sure if the name is right. I'm considering the possibility that we may want to build browser features with WebGPU in the future (like a compositor, or filters). Those wouldn't be tied to canvas, and presumably would instead have their own subclass of WebGPUSurface, right? Both subclasses would be compiled on Cocoa, though.
6. Is it really right that there's no concept of swap chain here? The GPUP doesn't seem to have any notion of this Surface object either, which I think is kind of contrary to the API design of WebGPU.framework. I kind of think we should try to have WebGPU.framework have a single usage pattern here for both CAMetalLayer and for WebKit, and try to make them as conceptually similar as possible. The current patch seems to be sidestepping this design by creating a new API-exposed WGPUTextureDescriptorIOSurfaceBacking object to create a texture directly from an IOSurface, without using the Surface/Swapchain machinery. I kind of think that's backwards.

Reply:
I'm going to break this part out into a separate PR